File: c++/cote/GJM/최단_경로_계산.cpp
#include <iostream>
#include <random>
#include <vector>
#include <queue>
#include <functional>
#define N 1000
#define TEST_CASE 10

using namespace std;

vector<vector<int>> initialize(int n, function<bool()> generator) {
	vector<vector<int>> ret(n, vector<int>{});

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < i ; j++) {
			if (generator()) {
				continue;
			}
			ret[i].push_back(j);
			ret[j].push_back(i);
		}
	}

	return ret;
}

vector<int> bfs(int start, vector<vector<int>>& adj) {
	vector<int> parent(adj.size(), -1);
	vector<int> distance(adj.size(), -1);

	queue<int> q;
	distance[start] = 0;
	parent[start] = start;
	q.push(start);

	while (!q.empty()) {
		int cur = q.front();
		q.pop();

		for (const auto& next : adj[cur]) {
			if (distance[next] == -1) {
				q.push(next);
				distance[next] = distance[cur] + 1;
				parent[next] = cur;
			}
		}
	}

	return parent;
}

vector<int> shortest_path(int target, const vector<int>& parent) {
	vector<int> path(1, target);

	while(parent[target] != target) {
		if (parent[target] == -1) {
			return {};
		}

		target = parent[target];
		path.push_back(target);
	}

	reverse(begin(path), end(path));
	return path;
}

vector<int> solve(int start, int target, vector<vector<int>>& adj) {
	vector<int> parent = bfs(start, adj);

	return shortest_path(target, parent);
}

int main(void) {
	default_random_engine gen(43);
	auto generator {bind(bernoulli_distribution {0.9775}, gen)};

	for (int it = 0; it < TEST_CASE; it++) {
		int n = gen() % (N - 3) + 3;
		vector<vector<int>> adj = initialize(n, generator);
		int k = gen() % (n - 1) + 1;
		vector<int> path = solve(0, k, adj);

		for (auto& cur : path) {
			cout << cur << " ";
		}
		cout << endl;
	}

	return 0;
}

File: go/cote/GJM/최단_경로_계산.go
package main

import (
	"fmt"
	"math/rand"
	"container/list"
	"slices"
)

const (
	TEST_CASES = 10
	N = 1000
)

func initialize(random *rand.Rand) ([][]int, int, int) {
	n := random.Intn(N - 3) + 3
	ret := make([][]int, n)

	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			if random.Float32() < 0.9775 {
				continue
			}
			ret[i] = append(ret[i], j)
			ret[j] = append(ret[j], i)
		}
	}

	a := random.Intn(n - 1) + 1
	b := random.Intn(a)
	return ret, b, a
}

func bfs(src int, adj [][]int) []int {
	distance := make([]int, len(adj))
	parent := make([]int, len(adj))

	for i, _ := range distance {
		distance[i] = -1
	}

	for i, _ := range parent {
		parent[i] = -1
	}

	queue := list.New()
	queue.PushBack(src)
	parent[src] = src 
	distance[src] = 0

	for{
		if queue.Len() == 0 {
			break
		}

		cur := queue.Front().Value.(int)
		queue.Remove(queue.Front())

		for _, next := range adj[cur] {
			if distance[next] == -1 {
				queue.PushBack(next)
				parent[next] = cur
				distance[next] = distance[cur] + 1
			}
		}
	}

	return parent
}

func shoretest_path(target int, parent []int) []int {
	ret := make([]int, 0)

	ret = append(ret, target)
	for {
		if parent[target] == - 1 {
			return make([]int, 0)
		}

		if parent[target] == target {
			break
		}
		ret = append(ret, parent[target])
		target = parent[target]
	}

	slices.Reverse(ret)

	return ret
}

func solve(src, target int, adj [][]int) []int {
	parent := bfs(src, adj)

	return shoretest_path(target, parent)
}

func main() {
	random := rand.New(rand.NewSource(43))
	for it := 0; it < TEST_CASES; it++ {
		adj, src, target := initialize(random)
		fmt.Println(solve(src, target, adj))
	}
}